### PR TITLE
p2p: filter peers by IP address and ID

### PR DESCRIPTION
**Pull request by tychoish**
Wanted to validate the approach before going to far down. 

I think the real risk here is that an application can put back pressure on the p2p layer which might not be desirable.

### CODE REVIEW

File: p2p/router.go [hunk 1]
```diff
@@ -507,6 +525,13 @@ func (r *Router) acceptPeers(transport Transport) {
 			defer conn.Close()
 			defer r.connTracker.RemoveConn(incomingIP)
 
+			if err := r.filterPeersIP(ctx, incomingIP, re.Port); err != nil {
```
[the remark this answers is not in the document]
Comment (tychoish) [reply p2p/router.go:528]: `RemoteAddr` is a method on `net.Conn` which the switch gets at by type-casting the `p2p.Connection` object to the `mConnection` implementation and then accessing a private field. 

This implementation avoids a direct dependency on the transport, which I think seems desirable.

[the remark this answers is not in the document]
Comment (alexanderbez) [reply p2p/router.go:528]: Makes sense 👍

File: p2p/router.go [hunk 2]
```diff
@@ -139,6 +140,9 @@ type RouterOptions struct {
 	// can attempt to create a new connection. Defaults to 10
 	// milliseconds, and cannot be less than 1 millisecond.
 	IncomingConnectionWindow time.Duration
+
+	FilterPeerByIP func(context.Context, net.IP, uint16) error
+	FilterPeerByID func(context.Context, NodeID) error
```
Comment (alexanderbez) [inline p2p/router.go:145]: Can we add some short documentation for both of these?

**Review by alexanderbez — APPROVED** [review]
Looks great!

**Pull request merged**